--- ltmain.sh.orig	2022-12-28 22:52:28.778793502 +0800
+++ ltmain.sh	2022-12-28 22:52:35.234686855 +0800
@@ -9703,20 +9703,37 @@
 
 	if test -n "$soname_spec"; then
 	  eval soname=\"$soname_spec\"
 	else
 	  soname=$realname
 	fi
 	if test -z "$dlname"; then
 	  dlname=$soname
 	fi
 
+## SANE modification (start)
+## Moved here from just before the enclosing 'fi' statement so we can
+## safely rewrite the soname before anything gets run.
+	# If -module or -export-dynamic was specified, set the dlname.
+	if test "$module" = yes || test "$export_dynamic" = yes; then
+	  # On all known operating systems, these are identical.
+	  dlname="$soname"
+	fi
+## The SANE API recommends/requires(?) that backends are made available
+## with the soname "libsane" instead of "libsane-$backend".  The dlname
+## should contain the -$backend part for libsane-dll to work.
+	case $host in
+	  *)
+	    soname=`echo $soname | sed -e "s/libsane-[A-Za-z_0-9]*/libsane/g"`
+	esac
+## SANE modification (end)
+
 	lib=$output_objdir/$realname
 	linknames=
 	for link
 	do
 	  func_append linknames " $link"
 	done
 
 	# Use standard objects if they are pic
 	test -z "$pic_flag" && libobjs=`$ECHO "$libobjs" | $SP2NL | $SED "$lo2o" | $NL2SP`
 	test "X$libobjs" = "X " && libobjs=
@@ -10162,25 +10179,28 @@
 	  exit $EXIT_SUCCESS
 	fi
 
 	# Create links to the real library.
 	for linkname in $linknames; do
 	  if test "$realname" != "$linkname"; then
 	    func_show_eval '(cd "$output_objdir" && $RM "$linkname" && $LN_S "$realname" "$linkname")' 'exit $?'
 	  fi
 	done
 
-	# If -module or -export-dynamic was specified, set the dlname.
-	if test yes = "$module" || test yes = "$export_dynamic"; then
-	  # On all known operating systems, these are identical.
-	  dlname=$soname
-	fi
+## SANE modification (start)
+## Moved before the soname modification needed by SANE.
+##	# If -module or -export-dynamic was specified, set the dlname.
+##	if test "$module" = yes || test "$export_dynamic" = yes; then
+##	  # On all known operating systems, these are identical.
+##	  dlname="$soname"
+##	fi
+## SANE modification (end)
       fi
       ;;
 
     obj)
       if test -n "$dlfiles$dlprefiles" || test no != "$dlself"; then
 	func_warning "'-dlopen' is ignored for objects"
       fi
 
       case " $deplibs" in
       *\ -l* | *\ -L*)
